### PR TITLE
feat: store deployed processes by id and version tag

### LINKED ISSUE #21035: Store deployed processes by id and version tag
### Description
To support the new binding type `versionTag` for call activities, Zeebe needs to be able to resolve a called process by process id and version tag.

### Acceptance Criteria
- [x] When a BPMN process is deployed, the (optional) version tag is read from the XML (`zeebe:versionTag` extension element) and stored in a new property `versionTag` in the created `ProcessRecord` and `PersistedProcess`.
- [x] A new column family is added to store a reference to the respective `processDefinitionKey` by `processId` and `versionTag`.
- [x] A new method is added to the `ProcessState` that allows to find a deployed process by `processId` and `versionTag`.

### Dependencies
- requires https://github.com/camunda/camunda/issues/21124
- required by #21038

### PR DESCRIPTION
**Pull request by jfriedenstab**
## Description
Store the version tag as a new property for deployed processes and add the possibility to find a process by process id and version tag. This is a prerequisite to be able to properly resolve called processes when the new `versionTag` binding type is used (see #21038).

## Related issues
closes #21035

### CODE REVIEW

File: zeebe/exporters/elasticsearch-exporter/src/main/resources/zeebe-record-process-template.json [hunk 1]
```diff
@@ -45,6 +45,9 @@
             },
             "deploymentKey": {
               "type": "long"
+            },
+            "versionTag": {
+              "type": "keyword"
```
Comment (jfriedenstab) [inline zeebe/exporters/elasticsearch-exporter/src/main/resources/zeebe-record-process-template.json:50]: FYI: I was a bit unsure if I should use type `keyword` or `text`, but from https://www.elastic.co/guide/en/elasticsearch/reference/current/keyword.html it sounded like `keyword` would be more appropriate:

> used for structured content such as IDs, email addresses, hostnames, status codes, zip codes, or **tags**.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbProcessState.java [hunk 2]
```diff
@@ -94,6 +94,19 @@ public final class DbProcessState implements MutableProcessState {
...
       processDefinitionKeyByProcessIdAndDeploymentKeyColumnFamily;
 
+  private final DbString versionTag;
+  private final DbTenantAwareKey<DbCompositeKey<DbString, DbString>>
+      tenantAwareProcessIdAndVersionTagKey;
+
+  /**
+   * <b>Note</b>: Will only be filled with processes deployed from 8.6 onwards that have a version
+   * tag assigned (which is an optional property).
+   */
```
Comment (koevskinikola) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbProcessState.java:104]: ❓ Does this mean that we don't need to perform a migration on already deployed processes?

🤔 I wonder if it would be good to have a migration test to confirm that there are no migration issues due to this change.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/resource/ResourceDeletionDeleteProcessor.java
Comment (koevskinikola) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/resource/ResourceDeletionDeleteProcessor.java:1]: 🔧 It would be good to put the resource deletion changes in a separate commit for better readability of the changes.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/resource/ResourceDeletionTest.java
Comment (koevskinikola) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/resource/ResourceDeletionTest.java:1]: 🔧 It would be good to put the resource deletion changes in a separate commit for better readability of the changes.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessCreatedV2Applier.java [hunk 3]
```diff
@@ -25,5 +25,6 @@ public ProcessCreatedV2Applier(final MutableProcessingState state) {
   public void applyState(final long processDefinitionKey, final ProcessRecord value) {
     processState.putProcess(processDefinitionKey, value);
     processState.storeProcessDefinitionKeyByProcessIdAndDeploymentKey(value);
+    processState.storeProcessDefinitionKeyByProcessIdAndVersionTag(value);
```
Comment (koevskinikola) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessCreatedV2Applier.java:28]: 💭 Just commenting for clarity:

This class was added with PR #20498 so any records that are processed with this applier must be created by Zeebe `8.6.0`, i.e. they will be `versionTag`-aware.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbProcessState.java [hunk 4]
```diff
@@ -231,6 +269,10 @@ public void deleteProcess(final ProcessRecord processRecord) {
     processDefinitionKeyByProcessIdAndDeploymentKeyColumnFamily.deleteIfExists(
         tenantAwareProcessIdAndDeploymentKey);
 
+    // use deleteIfExists as no entry exists for processes that do not have a version tag assigned
+    processDefinitionKeyByProcessIdAndVersionTagColumnFamily.deleteIfExists(
+        tenantAwareProcessIdAndVersionTagKey);
+
```
Comment (koevskinikola) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbProcessState.java:275]: 🔧 It would be good to put the process deletion changes in a separate commit for better readability.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbProcessState.java [hunk 5]
```diff
@@ -372,6 +414,22 @@ public DeployedProcess getProcessByProcessIdAndVersion(
...
 
+  @Override
+  public DeployedProcess getProcessByProcessIdAndVersionTag(
+      final DirectBuffer processId, final String versionTag, final String tenantId) {
+    tenantIdKey.wrapString(tenantId);
+    this.processId.wrapBuffer(processId);
+    this.versionTag.wrapString(versionTag);
+    final var foreignKey =
+        processDefinitionKeyByProcessIdAndVersionTagColumnFamily.get(
+            tenantAwareProcessIdAndVersionTagKey);
```
[the remark this answers is not in the document]
Comment (koevskinikola) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbProcessState.java:425]: I agree, I'm not sure how cache invalidation would work. All the process caches until now add a unique combination (ex. tenant and key, process id and version, etc) that may push out an older entry from the cache, but not invalidate it.

Maybe you can create a follow-up "enhancement" issue for this? That way, if we notice performance issues with the look-up by process id and version tag, we can add them to that issue and prioritize it.

**Review by koevskinikola — CHANGES_REQUESTED** [review]
Nice job @jfriedenstab! 👏 

🙃 There are a lot of comments, but most of them are nitpicks for improving the readability of this PR or the code in some of the classes.

:x: I noted some (probably) required changes for the Tasklist classes.

❓ I'm also wondering if a migration test would be useful here, to ensure that no existing deployed processes are impacted by this change.

**Comment by koevskinikola** [comment]
🔧 @jfriedenstab the `MultiPartitionDeploymentLifecycleTest` is also failing, it might be related to the changes.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbProcessState.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jfriedenstab) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbProcessState.java:425]: Thanks, good suggestion! 👍🏻 I'll create an issue.

File: tasklist/importer-850/src/main/java/io/camunda/tasklist/zeebeimport/v850/record/value/deployment/DeployedProcessImpl.java
[the remark this answers is not in the document]
Comment (jfriedenstab) [reply tasklist/importer-850/src/main/java/io/camunda/tasklist/zeebeimport/v850/record/value/deployment/DeployedProcessImpl.java:1]: Ah, thanks, I totally missed those methods 🤦🏻.

**Comment by jfriedenstab** [comment]
> `MultiPartitionDeploymentLifecycleTest` is also failing, it might be related to the changes.

My assumption is that it's because of the new `versionTag` property in the record that is being written (which makes the record slightly bigger and causes an `EXCEEDED_BATCH_RECORD_SIZE` error). It can be fixed by generating a slightly smaller resource in `shouldDeployIfResourceIsLargeButNotTooMuch`.

File: zeebe/exporters/elasticsearch-exporter/src/main/resources/zeebe-record-deployment-template.json
[the remark this answers is not in the document]
Comment (jfriedenstab) [reply zeebe/exporters/elasticsearch-exporter/src/main/resources/zeebe-record-deployment-template.json:1]: Good suggestion! I've split the commits.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbProcessState.java (hunk 2 above)
Comment (jfriedenstab) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbProcessState.java:104]: I'm not really familiar with the migrations, but in my opinion, we don't need to perform one, because already deployed processes won't have a version tag anyway.

By the way, I also had a similar discussion with Nico in the context of "deployment version binding" and we concluded that a migration would not be necessary (see https://github.com/camunda/camunda/issues/19709#issuecomment-2245723417 and the comments above).

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessCreatedV2Applier.java (hunk 3 above)
Comment (jfriedenstab) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessCreatedV2Applier.java:28]: Yes, I figured that it would be a similar case as above where processes are stored by id and deployment key.

**Comment by jfriedenstab** [comment]
@koevskinikola Thanks a lot for the review! I've addressed all your comments.

**Comment by jfriedenstab** [comment]
@camunda/hto: I needed to make some changes in the `tasklist` module (see `DeployedProcessImpl` classes) due to a new field in the `ProcessMetadataValue` interface, maybe one of you could have a quick look?

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbProcessState.java (hunk 2 above)
Comment (koevskinikola) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbProcessState.java:104]: Ah okay, that's good to see. Then consider this discussion resolved 😄

**Review by koevskinikola — APPROVED** [review]
All good from my side @jfriedenstab! 🚀

**Review by marcosgvieira — APPROVED** [review]
Everything looking good from Tasklist perspective, thanks @jfriedenstab
